File: java/examples/cc0mon-api-get-collector.java
///usr/bin/env jbang "$0" "$@" ; exit $?
//JAVA 17
//DEPS com.cc0mon:sdk:0.1.0
//DEPS com.fasterxml.jackson.core:jackson-databind:2.17.0

import com.cc0mon.sdk.Client;
import com.cc0mon.sdk.Errors;
import com.cc0mon.sdk.LoggingSetup;
import com.cc0mon.sdk.Models;
import com.cc0mon.sdk.Models.Collector;
import com.cc0mon.sdk.Models.CollectorItem;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.SerializationFeature;

import java.util.List;
import java.util.logging.Logger;

class GetCollector {
    public static void main(String[] args) {
        Logger log = LoggingSetup.configure("get-collector");
        String address = null;
        String energy = null, rarity = null;
        boolean ownedOnly = false;
        for (int i = 0; i < args.length; i++) {
            switch (args[i]) {
                case "--address": if (i + 1 < args.length) address = args[++i]; break;
                case "--energy": if (i + 1 < args.length) energy = args[++i]; break;
                case "--rarity": if (i + 1 < args.length) rarity = args[++i]; break;
                case "--owned-only": ownedOnly = true; break;
                case "-h": case "--help":
                    System.out.println("usage: cc0mon-api-get-collector.java --address 0x... [--owned-only] [--energy E] [--rarity R]");
                    System.out.println("Valid --energy values: " + String.join(", ", Models.ENERGIES));
                    System.out.println("Valid --rarity values: " + String.join(", ", Models.RARITIES));
                    System.exit(0);
            }
        }
        if (address == null) {
            System.err.println("usage: cc0mon-api-get-collector.java --address 0x... [--owned-only] [--energy E] [--rarity R]");
            System.exit(5);
        }
        boolean hasFilter = ownedOnly || energy != null || rarity != null;
        log.info("script start address=" + address + " owned_only=" + ownedOnly + " energy=" + energy + " rarity=" + rarity);

        try (Client client = new Client()) {
            ObjectMapper mapper = new ObjectMapper().enable(SerializationFeature.INDENT_OUTPUT);
            if (hasFilter) {
                List<CollectorItem> items = client.findCollectorItems(address, ownedOnly, energy, rarity);
                System.out.println(mapper.writeValueAsString(items));
                log.info("script exit code=0 matched=" + items.size());
            } else {
                Collector c = client.getCollector(address);
                System.out.println(mapper.writeValueAsString(c));
                log.info("script exit code=0 collected=" + c.collected() + " missing=" + c.missing() + " total_tokens_held=" + c.totalTokensHeld());
            }
        } catch (Errors.ValidationException e) { log.severe("validation: " + e.getMessage()); System.err.println("validation error: " + e.getMessage()); System.exit(5);
        } catch (Errors.NetworkException e)    { log.severe("network: " + e.getMessage());    System.err.println("network error: " + e.getMessage()); System.exit(2);
        } catch (Errors.ClientApiException e)  { log.severe("http " + e.getStatusCode());     System.err.println("client error HTTP " + e.getStatusCode() + ": " + e.getBody()); System.exit(3);
        } catch (Errors.ServerApiException e)  { log.severe("http " + e.getStatusCode());     System.err.println("server error HTTP " + e.getStatusCode() + ": " + e.getBody()); System.exit(4);
        } catch (Exception e) { log.severe("unexpected: " + e); System.err.println("unexpected: " + e); System.exit(1); }
    }
}

File: python/examples/cc0mon-api-get-collector.py
"""cc0mon-api-get-collector — fetch a wallet's collection checklist.

Usage::

    python cc0mon-api-get-collector.py --address 0xabc...def

Only 0x-prefixed hex addresses are accepted (no ENS).
"""
from __future__ import annotations

import argparse
import dataclasses
import json
import sys

from cc0mon_sdk import (
    Client,
    ClientApiError,
    NetworkError,
    ServerApiError,
    ValidationError,
    configure_logger,
)

ACTION = "get-collector"


def main(argv: list[str] | None = None) -> int:
    parser = argparse.ArgumentParser(description="Fetch a wallet's cc0mon collection checklist.")
    parser.add_argument("--address", required=True, help="Ethereum 0x-address (42 chars)")
    args = parser.parse_args(argv)

    logger = configure_logger(ACTION)
    logger.info("script start address=%s", args.address)

    try:
        with Client() as client:
            collector = client.get_collector(args.address)
    except ValidationError as e:
        logger.error("validation: %s", e)
        print(f"validation error: {e}", file=sys.stderr)
        return 5
    except NetworkError as e:
        logger.error("network: %s", e)
        print(f"network error: {e}", file=sys.stderr)
        return 2
    except ClientApiError as e:
        logger.error("http %d: %s", e.status_code, e.body[:200])
        print(f"client error HTTP {e.status_code}: {e.body[:200]}", file=sys.stderr)
        return 3
    except ServerApiError as e:
        logger.error("http %d: %s", e.status_code, e.body[:200])
        print(f"server error HTTP {e.status_code}: {e.body[:200]}", file=sys.stderr)
        return 4

    print(json.dumps(dataclasses.asdict(collector), indent=2, default=str))
    logger.info(
        "script exit code=0 progress_keys=%d items=%d",
        len(collector.progress),
        len(collector.items),
    )
    return 0


if __name__ == "__main__":
    sys.exit(main())
